pkg/unpack/unpacker.go

```diff
@@ -47,48 +48,170 @@ const (
 	labelSnapshotRef = "containerd.io/snapshot.ref"
 )
 
+// Unpacker unpacks images by hooking into the image handler process.
+// Unpacks happen in the backgrounds and waited on to complete.
+type Unpacker interface {
+	Unpack(images.Handler) images.Handler
+	Wait() (Result, error)
+}
+
+// Result returns information about the unpacks which were completed.
+type Result struct {
+	Unpacks int
+}
+
+type unpackerConfig struct {
+	platforms []*Platform
+
+	content content.Store
+
+	limiter               *semaphore.Weighted
+	duplicationSuppressor kmutex.KeyedLocker
+}
+
+type Platform struct {
+	Platform platforms.Matcher
+
+	SnapshotterName string
+	Snapshotter     snapshots.Snapshotter
+	SnapshotOpts    []snapshots.Opt
+
+	Applier   diff.Applier
+	ApplyOpts []diff.ApplyOpt
+}
+
+type UnpackerOpt func(*unpackerConfig) error
+
+func WithUnpackPlatform(u Platform) UnpackerOpt {
+	return UnpackerOpt(func(c *unpackerConfig) error {
+		if u.Platform == nil {
+			u.Platform = platforms.All
+		}
+		if u.SnapshotterName == "" {
+			return fmt.Errorf("snapshotter name must be provided for reference")
+		}
+		if u.Snapshotter == nil {
+			return fmt.Errorf("snapshotter must be provided to unpack")
+		}
```

Comment:
How does that work if SnapshotterName and Snapshotter are different?

Reply:
Would the caller be responsible to set them in the consistent manner?

Reply:
Could rename `SnapshotterName` to `SnapshotterKey` and skip the gc reference when not provided. We probably should have been adding a `Name() string` or to some of our interfaces or rely on checking for `String() string` so callers are forced to track the name/object pairing.

Reply:
Renamed `SnapshotterName` and removed it erroring. The key only needs to be consistent for the caller, but it doesn't need to cause an error.